### DISCUSSION TITLE
Why does MutableValue not emit distinct values?

### DISCUSSION THREAD

**Discussion by eosobande** [arkivanov/Decompose#755] [Q&A]
```
val distinct = MutableValue(0)

distinct.subscribe { println("SUB: $it") }

distinct.update { 0 }

// RESULT
// SUB: 0
// SUB: 0
// prints twice, one for the initial subscription and again for the updated value
```

`MutableValue::compareAndSet` does not check if the value is changing or not, it just sets the new value and emits.

Proposed FIX in `MutableValue::setValue`:
```
private inline fun setValue(value: T, predicate: (T) -> Boolean = { true }): Boolean {
        lock.synchronized {
            if (!predicate(_value)) {
                return false
            }

            if (_value == value) return true // Don't do anything if value is not changing

            _value = value

            if (isEmitting) {
                return true
            }

            isEmitting = true
        }

        emit()

        return true
}
```

**Comment by arkivanov** [comment]
That's because `MutableValue` is similar to `BehaviorSubject` from RxJava. It's not necessarily a state holder, but rather an arbitrary value holder. Also, the proposed change would be a behaviour change, which is only possible with a major release (e.g. `4.0.0`).

This however shouldn't not be an issue, since UI frameworks should be handling this case just fine. E.g. with Jetpack Compose, `MutableValue` is observed via `mutableStateOf`.

**Comment by eosobande** [comment]
Thanks @arkivanov that answers my question
